solr/core/src/java/org/apache/solr/metrics/MaxHistogram.java

```diff
@@ -0,0 +1,191 @@
+
+  private static class RelayClock extends Clock {
+
+    private long nextTick;
+
+    private RelayClock(long initialTick) {
+      nextTick = initialTick;
+    }
+
+    @Override
+    public long getTick() {
+      return nextTick;
+    }
+
+    public void setTick(long value) {
+      nextTick = value;
+    }
+  }
+
+  /**
+   * Constructs a new {@link MaxHistogram} instance.
+   *
+   * @param intervalMillis this is the unit interval at which values will be recorded into the
+   *     underlying histogram (the max value for the immediately preceding interval). This
+   *     granularity is somewhat arbitrary, and the default of 1000ms (1s) should be sufficient for
+   *     most applications.
+   * @param maxBackdateSeconds limits how far back value updates will be lazily backfilled upon
+   *     value change or read access. Without this, the first value change or read access after a
+   *     long period of inactivity could result in backfilling an arbitrarily large number of
+   *     intervals. This should ordinarily be set to align with the time window of the reservoir (or
+   *     300s/5m for the default configuration of {@link
+   *     com.codahale.metrics.ExponentiallyDecayingReservoir}).
+   * @param clock the backing clock to use for the reservoir
+   * @param reservoirFunction because {@link MaxHistogram} internally shims the clock actually used
+   *     by the underlying reservoir, the reservoir cannot be constructed directly. This function
+   *     supports constructing a reservoir based on the shimmed clock
+   * @return a new {@link MaxHistogram} instance
+   */
+  public static MaxHistogram newInstance(
+      int intervalMillis,
+      int maxBackdateSeconds,
+      Clock clock,
+      Function<Clock, Reservoir> reservoirFunction) {
+    RelayClock relayClock = new RelayClock(clock.getTick());
+    return new MaxHistogram(
+        intervalMillis, maxBackdateSeconds, reservoirFunction.apply(relayClock), relayClock, clock);
+  }
+
+  private MaxHistogram(
+      int intervalMillis,
+      int maxBackdateSeconds,
+      Reservoir reservoir,
+      Clock relayClock,
+      Clock clock) {
+    super(reservoir);
+    this.intervalNanos = TimeUnit.MILLISECONDS.toNanos(intervalMillis);
+    if (maxBackdateSeconds == -1) {
+      this.maxBackdateNanos = Long.MAX_VALUE; // effectively no limit
+    } else if (TimeUnit.SECONDS.toMillis(maxBackdateSeconds) >= intervalMillis) {
+      // align `maxBackdateNanos` to a multiple of `intervalNanos`
+      long requestedMaxBackdate = TimeUnit.SECONDS.toNanos(maxBackdateSeconds);
+      this.maxBackdateNanos = requestedMaxBackdate - (requestedMaxBackdate % intervalNanos);
+    } else {
+      throw new IllegalArgumentException(
+          "maxBackdateSeconds must be -1 (unlimited) or >= "
+              + TimeUnit.MILLISECONDS.toSeconds(intervalMillis)
+              + "; found "
+              + maxBackdateSeconds);
+    }
+    this.relayClock = (RelayClock) relayClock;
```

Comment:
Why is the param not RelayClock then?

Reply:
💯 -- this was leftover from some refactoring and had to do with class scope/visibility ... no longer an issue, will fix.